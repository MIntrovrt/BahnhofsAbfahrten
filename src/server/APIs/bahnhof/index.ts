import { getLageplan } from 'server/Bahnhof/Lageplan';
import KoaRouter from 'koa-router';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter().get('/lageplan/:stationName', async ctx => {
    const { stationName }: { stationName: string } = ctx.params;

    ctx.body = {
      lageplan: await getLageplan(stationName),
    };
  });

router
  .prefix('/bahnhof')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes());

export default router;
export const versions = ['current', 'v1'];
