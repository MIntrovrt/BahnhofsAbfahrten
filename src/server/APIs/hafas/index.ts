import auslastungHafas from 'server/Auslastung/Hafas';
import geoStation from 'server/HAFAS/LocGeoPos';
import journeyDetails from 'server/HAFAS/JourneyDetails';
import KoaRouter from 'koa-router';
import LocMatch from 'server/HAFAS/LocMatch';
import makeRequest from 'server/HAFAS/Request';
import routing from 'server/HAFAS/TripSearch';
import stationBoard from 'server/HAFAS/StationBoard';
import trainSearch from 'server/HAFAS/TrainSearch';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter()
    .get('/journeyDetails/:jid', async ctx => {
      const { jid }: { jid: string } = ctx.params;

      ctx.body = await journeyDetails(jid);
    })
    .get('/details/:trainName/:date', async ctx => {
      const { date, trainName } = ctx.params;

      const trains = await trainSearch(trainName, Number.parseInt(date, 10));

      if (trains.length) {
        ctx.body = await journeyDetails(trains[0].jid);
      } else {
        ctx.status = 404;
      }
    })
    .get('/auslastung/:start/:dest/:trainNumber/:time', async ctx => {
      const { start, dest, time, trainNumber } = ctx.params;

      ctx.body = await auslastungHafas(
        start,
        dest,
        trainNumber,
        Number.parseInt(time, 10)
      );
    })
    .get('/ArrStationBoard', async ctx => {
      const { date, station } = ctx.query;

      ctx.body = await stationBoard({
        type: 'ARR',
        station,
        date: Number.parseInt(date, 10) || undefined,
      });
    })
    .get('/DepStationBoard', async ctx => {
      const { date, station } = ctx.query;

      ctx.body = await stationBoard({
        type: 'DEP',
        station,
        date: Number.parseInt(date, 10) || undefined,
      });
    })
    .get('/trainSearch/:trainName/:date', async ctx => {
      const { date, trainName } = ctx.params;

      ctx.body = await trainSearch(trainName, Number.parseInt(date, 10));
    })
    .get('/geoStation', async ctx => {
      const { x, y, lat, lng, maxDist = 1000 } = ctx.query;

      const realY = lat ? Number.parseFloat(lat) * 1000000 : y;
      const realX = lng ? Number.parseFloat(lng) * 1000000 : x;

      ctx.body = await geoStation(realX, realY, maxDist);
    })
    .get('/station/:searchTerm', async ctx => {
      const { searchTerm } = ctx.params;

      ctx.body = await LocMatch(searchTerm);
    })
    .post('/rawHafas', async ctx => {
      ctx.body = await makeRequest(ctx.request.body);
    })
    .post('/route', async ctx => {
      ctx.body = await routing({
        ...ctx.request.body,
        time: Number.parseInt(ctx.request.body.time, 10),
      });
    });

router
  .prefix('/hafas')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes());

export default router;
export const versions = ['v1'];
