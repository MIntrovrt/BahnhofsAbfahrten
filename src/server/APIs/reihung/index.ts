import {
  wagenReihung,
  wagenReihungMonitoring,
  wagenReihungStation,
} from 'server/Reihung';
import KoaRouter from 'koa-router';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter()
    .get('/wagenstation/:train/:station', async ctx => {
      const { train, station }: { train: string; station: string } = ctx.params;

      ctx.body = await wagenReihungStation([train], station);
    })
    .get('/wagen/:trainNumber/:date', async ctx => {
      const {
        date,
        trainNumber,
      }: { date: string; trainNumber: string } = ctx.params;

      ctx.body = await wagenReihung(trainNumber, Number.parseInt(date, 10));
    });

router
  .prefix('/reihung')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes())
  .get('/monitoring/wagen', async ctx => {
    ctx.body = await wagenReihungMonitoring();
    if (!ctx.body) {
      ctx.status = 404;
    }
  });

export default router;
export const versions = ['v1'];
