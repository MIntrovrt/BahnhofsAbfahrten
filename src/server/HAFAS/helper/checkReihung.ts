import { TrnCmpSX } from 'types/HAFAS';

const allowdTypes = ['ICE', 'IC', 'TGV', 'EC', 'ECE', 'RJ', 'D'];

export default (trnCmpSX?: TrnCmpSX, trainType?: string) => {
  if (trainType && allowdTypes.includes(trainType)) return true;

  if (trnCmpSX && trnCmpSX.tcM) return true;
};
