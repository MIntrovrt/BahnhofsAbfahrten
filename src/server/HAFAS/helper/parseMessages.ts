import { MsgL, ParsedCommon } from 'types/HAFAS';

export default (msgL: MsgL[], common: ParsedCommon) => {
  const messages = msgL.map(msg => common.remL[msg.remX]);

  return messages;
};
