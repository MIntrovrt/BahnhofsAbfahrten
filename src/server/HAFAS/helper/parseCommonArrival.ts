import { CommonArrival, ParsedCommon } from 'types/HAFAS';
import { differenceInMinutes } from 'date-fns';
import { ParsedCommonArrival } from 'types/common';
import parseMessages from './parseMessages';
import parseTime from './parseTime';

export default (
  a: CommonArrival,
  date: number,
  common: ParsedCommon,
  trainType?: string
): ParsedCommonArrival => {
  const scheduledTime = parseTime(date, a.aTimeS);
  let time = scheduledTime;
  let delay;

  if (a.aTimeR) {
    time = parseTime(date, a.aTimeR);
    delay = time && scheduledTime && differenceInMinutes(time, scheduledTime);
  }

  return {
    scheduledPlatform: a.aPlatfR && a.aPlatfS,
    platform: a.aPlatfR || a.aPlatfS,
    scheduledTime,
    time,
    delay,
    reihung:
      Boolean(a.aTrnCmpSX && a.aTrnCmpSX.tcM) ||
      trainType === 'IC' ||
      trainType === 'EC' ||
      undefined,
    messages: a.msgL ? parseMessages(a.msgL, common) : undefined,
  };
};
