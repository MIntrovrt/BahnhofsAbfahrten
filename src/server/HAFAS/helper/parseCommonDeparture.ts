import { CommonDeparture, ParsedCommon } from 'types/HAFAS';
import { differenceInMinutes } from 'date-fns';
import { ParsedCommonDeparture } from 'types/common';
import parseMessages from './parseMessages';
import parseTime from './parseTime';

export default (
  d: CommonDeparture,
  date: number,
  common: ParsedCommon,
  trainType?: string
): ParsedCommonDeparture => {
  const scheduledTime = parseTime(date, d.dTimeS);
  let time = scheduledTime;
  let delay;

  if (d.dTimeR) {
    time = parseTime(date, d.dTimeR);
    delay = time && scheduledTime && differenceInMinutes(time, scheduledTime);
  }

  return {
    scheduledPlatform: d.dPlatfR && d.dPlatfS,
    platform: d.dPlatfR || d.dPlatfS,
    scheduledTime,
    time,
    delay,
    reihung:
      Boolean(d.dTrnCmpSX && d.dTrnCmpSX.tcM) ||
      trainType === 'IC' ||
      trainType === 'EC' ||
      undefined,
    messages: d.msgL ? parseMessages(d.msgL, common) : undefined,
  };
};
