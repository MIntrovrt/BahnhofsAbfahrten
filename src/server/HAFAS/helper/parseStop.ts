import { CommonStop, ParsedCommon } from 'types/HAFAS';
import { Route$Stop } from 'types/routing';
import parseAuslastung from './parseAuslastung';
import parseCommonArrival from './parseCommonArrival';
import parseCommonDeparture from './parseCommonDeparture';

export default (
  stop: CommonStop,
  common: ParsedCommon,
  date: number,
  trainType?: string
): Route$Stop => ({
  station: common.locL[stop.locX],
  arrival: stop.aTimeS
    ? parseCommonArrival(stop, date, common, trainType)
    : undefined,
  departure: stop.dTimeS
    ? parseCommonDeparture(stop, date, common, trainType)
    : undefined,
  auslastung: parseAuslastung(stop.dTrnCmpSX, common.tcocL),
});
