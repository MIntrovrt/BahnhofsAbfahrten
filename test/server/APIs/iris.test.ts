import { checkApi, mockWithFile } from './helper';
import { versions } from 'server/APIs/iris';
jest.mock('node-cache');

describe('IRIS API', () => {
  versions.forEach(v => {
    describe(v, () => {
      it('/abfahrten', async () => {
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8098105',
          'iris/station/8098105',
          'get',
          3
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8000105',
          'iris/station/8000105',
          'get',
          3
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8089211',
          'iris/station/8089211',
          'get',
          3
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/fchg/8098105',
          'iris/fchg/8098105'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/fchg/8000105',
          'iris/fchg/8000105'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/fchg/8089211',
          'iris/fchg/8089211'
        );
        mockWithFile(
          'https://www.bahnhof.de',
          '/service/search/bahnhof-de/520608?query=Frankfurt%20Hbf%20%20%28tief%29',
          'iris/bahnhof/frankfurt hbf tief'
        );
        mockWithFile(
          'https://www.bahnhof.de',
          '/service/search/bahnhof-de/520608?query=Frankfurt%20%28Main%29%20Hbf',
          'iris/bahnhof/frankfurt main hbf'
        );
        mockWithFile(
          'https://www.bahnhof.de',
          '/service/search/bahnhof-de/520608?query=Frankfurt%20%28M%29%20Hbf%20Stuttgarter%20Str.',
          'iris/bahnhof/frankfurt hbf stuttgarter str'
        );
        mockWithFile(
          'https://www.bahnhof.de',
          '/resource/blob/511998/2f507148493375276b906a182e429fbc/Zukunft-Bahn-DB-Station-Service-AG-data.pdf',
          'iris/bahnhof/zukunftdb.pdf'
        );
        mockWithFile(
          'https://www.bahnhof.de',
          '/bahnhof-de/bahnhof/Frankfurt__Main__Hbf-1038974',
          'iris/bahnhof/frankfurt main hbf result'
        );

        [15, 16, 17].forEach(time => {
          mockWithFile(
            'https://iris.noncd.db.de',
            `/iris-tts/timetable/plan/8098105/190627/${time}`,
            `iris/plan/8098105/${time}`
          );
          mockWithFile(
            'https://iris.noncd.db.de',
            `/iris-tts/timetable/plan/8000105/190627/${time}`,
            `iris/plan/8000105/${time}`
          );
          mockWithFile(
            'https://iris.noncd.db.de',
            `/iris-tts/timetable/plan/8089211/190627/${time}`,
            `iris/plan/8089211/${time}`
          );
        });

        await checkApi(
          `/api/iris/${v}/abfahrten/8098105?lookbehind=30&lookahead=30`
        );
      });
      it('/wings', async () => {
        mockWithFile(
          'https://iris.noncd.db.de',
          `/iris-tts/timetable/wingdef/5090299178847069383-1906271825-1/-688384725346268286-1906271825-1`,
          `iris/wingdef`
        );
        await checkApi(
          `/api/iris/${v}/wings/5090299178847069383-1906271825-1/-688384725346268286-1906271825-1`
        );
      });
    });
  });
});
