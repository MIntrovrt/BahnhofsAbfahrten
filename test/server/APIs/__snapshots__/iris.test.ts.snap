// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IRIS API v1 /abfahrten 1`] = `
Object {
  "departures": Array [],
  "lookbehind": Array [],
  "wings": Object {},
}
`;

exports[`IRIS API v1 /wings 1`] = `
Object {
  "end": Object {
    "fl": false,
    "pt": 1561658760000,
    "station": Object {
      "id": "8003040",
      "title": "Idar-Oberstein",
    },
  },
  "start": Object {
    "fl": true,
    "pt": 1561652700000,
    "station": Object {
      "id": "8000105",
      "title": "Frankfurt(Main)Hbf",
    },
  },
}
`;
